src/java/org/apache/cassandra/index/sai/disk/v2/sortedbytes/SortedBytesReader.java

```diff
@@ -0,0 +1,157 @@
+
+import static org.apache.cassandra.index.sai.disk.v2.sortedbytes.SortedBytesWriter.DIRECT_MONOTONIC_BLOCK_SHIFT;
+import static org.apache.cassandra.index.sai.disk.v2.sortedbytes.SortedBytesWriter.TERMS_DICT_BLOCK_MASK;
+import static org.apache.cassandra.index.sai.disk.v2.sortedbytes.SortedBytesWriter.TERMS_DICT_BLOCK_SHIFT;
+
+/**
+ * Lookup a point id by ByteComparable.
+ *   Implemented exactly from Lucene version 7.5 SortedDocValues
+ *
+ * Lookup a ByteComparable by point id.
+ *   Implemented using the trie with a long payload for the point id.
+ */
+public class SortedBytesReader
+{
+    private final SortedBytesWriter.Meta meta;
+    private final FileHandle trieHandle;
+    private final BytesRef term;
+    private final LongValues offsets;
+    private long pointId = -1;
+
+    public SortedBytesReader(SortedBytesWriter.Meta meta,
+                             FileHandle trieHandle,
+                             IndexInput offsetsInput) throws IOException
+    {
+        this.meta = meta;
+        this.trieHandle = trieHandle;
+
+        ByteArrayIndexInput offsestMetaInput = new ByteArrayIndexInput("", meta.offsetMetaBytes);
+
+        DirectMonotonicReader.Meta offsetsMeta = DirectMonotonicReader.loadMeta(offsestMetaInput, meta.offsetBlockCount, DIRECT_MONOTONIC_BLOCK_SHIFT);
+        RandomAccessInput offsetSlice = offsetsInput.randomAccessSlice(0, offsetsInput.length());
+
+        offsets = DirectMonotonicReader.getInstance(offsetsMeta, offsetSlice);
+
+        term = new BytesRef(meta.maxTermLength);
+    }
+
+    /**
+     * Returns the point id / ordinal of the target term, it not matching the next greater.
+     * @param term target term to lookup
+     * @return point id / ordinal or the target
+     * @throws IOException
+     */
+    public long seekToBytes(ByteComparable term) throws IOException
+    {
+        try (TrieRangeIterator reader = new TrieRangeIterator(trieHandle.instantiateRebufferer(),
+                                                              meta.trieFP,
+                                                              term,
+                                                              null,
+                                                              true,
+                                                              true))
+        {
+            final Iterator<Pair<ByteSource, Long>> iterator = reader.iterator();
+
+            if (iterator.hasNext())
+            {
+                final Pair<ByteSource, Long> pair = iterator.next();
+                return pair.right.longValue();
+            }
+            return -1;
+        }
+    }
+
+    /**
+     * Look up the ByteComparable of a target point id
+     *
+     * @param target Point id to lookup
+     * @param bytesInput Bytes input stream
+     * @return ByteComparable at the target point id
+     * @throws IOException
+     */
+    public ByteComparable seekExact(long target, IndexInput bytesInput) throws IOException
```

Comment:
I haven't read your comment and I also wondered, what was behind the decision that `bytesInput` is a parameter, while the other data structures like offsets are passed at the construction time. 

I'm not sure if it allows to share it between multiple threads, because `IndexInput` is not thread-safe anyways so you would not be able to share it. I'd rather see it as a way to decouple the lifetime of indexinput from that object and this way we can avoid additional reference counting, is it correct?